code/_onclick/hud/ai.dm

```diff
@@ -172,6 +172,65 @@
 	var/mob/living/silicon/ai/AI = usr
 	AI.drop_new_multicam()
 
+/atom/movable/screen/ai/floor_indicator
+	icon_state = "zindicator"
+	screen_loc = ui_ai_floor_indicator
+
+/atom/movable/screen/ai/floor_indicator/Initialize(mapload, datum/hud/hud_owner)
+	. = ..()
+	if(istype(hud_owner))
+		RegisterSignal(hud_owner, COMSIG_HUD_OFFSET_CHANGED, PROC_REF(update_z))
+
+/atom/movable/screen/ai/floor_indicator/proc/update_z(datum/hud/source)
+	SIGNAL_HANDLER
+	var/mob/living/silicon/ai/ai = source.mymob //if you use this for anyone else i will find you
+	if(isnull(ai))
+		return
+	var/turf/locturf = isturf(ai.loc) ? get_turf(ai.eyeobj) : get_turf(ai) //must be a var cuz error
+	var/ai_z = locturf.z
+	var/text = "Level<br/>[ai_z]"
+	if(SSmapping.level_trait(ai_z, ZTRAIT_STATION))
+		text = "Floor<br/>[ai_z - 1]"
+	else if (SSmapping.level_trait(ai_z, ZTRAIT_NOPHASE))
+		text = "ERROR"
+	maptext = MAPTEXT_TINY_UNICODE("<div align='center' valign='middle' style='position:relative; top:0px; left:0px'>[text]</div>")
+
+/atom/movable/screen/ai/go_up
+	name = "go up"
+	icon_state = "up"
+	screen_loc = ui_ai_godownup
+
+/atom/movable/screen/ai/go_up/Initialize(mapload)
+	. = ..()
+	register_context()
+
+/atom/movable/screen/ai/go_up/add_context(atom/source, list/context, obj/item/held_item, mob/user)
+	. = ..()
+	context[SCREENTIP_CONTEXT_LMB] = "Go up a floor"
+	return CONTEXTUAL_SCREENTIP_SET
+
+/atom/movable/screen/ai/go_up/Click(location,control,params)
+	var/mob/ai = get_mob() //the core
+	if(!isturf(ai.loc)) //aicard and stuff
+		return
+	usr.up()
```

Comment:
```suggestion
	ai.up()
```

there's no need to use `usr` here (even if it works) if we're already accessing the mob. just a safety thing in case `usr` and `ai` ever desync in REALLY strange ways.